File: ios/iCam/Connectivity/Security/SecureChannel.swift
import Foundation
import CryptoKit

/// The derived key material for one connection.
struct SessionKeys: Sendable, Equatable {
    var c2sKey: Data
    var s2cKey: Data
    var c2sSalt: Data      // 4 bytes
    var s2cSalt: Data      // 4 bytes
    /// Six digits shown on both devices during first pairing.
    var pairingDigits: String

    static func derive(sharedSecret: SharedSecret, clientRandom: Data, serverRandom: Data) -> SessionKeys {
        // CryptoKit's `SharedSecret` hashes nothing on its own, and .NET's
        // `DeriveRawSecretAgreement` returns the same raw X coordinate — which
        // is why both sides can reach an identical key schedule.
        let raw = sharedSecret.withUnsafeBytes { Data($0) }
        return derive(sharedSecretBytes: raw, clientRandom: clientRandom, serverRandom: serverRandom)
    }

    /// The same schedule, from raw bytes. Separated so the conformance vectors
    /// can pin it down without performing a key agreement.
    static func derive(sharedSecretBytes: Data, clientRandom: Data, serverRandom: Data) -> SessionKeys {
        var salt = Data()
        salt.append(clientRandom)
        salt.append(serverRandom)

        let ikm = SymmetricKey(data: sharedSecretBytes)
        let prk = HKDF<SHA256>.extract(inputKeyMaterial: ikm, salt: salt)

        func expand(_ info: String, _ count: Int) -> Data {
            let key = HKDF<SHA256>.expand(pseudoRandomKey: prk,
                                          info: Data(info.utf8),
                                          outputByteCount: count)
            return key.withUnsafeBytes { Data($0) }
        }

        let sasSeed = expand("iCam/v1 sas", 8)
        let value = sasSeed.readUInt64BE(at: 0) % 1_000_000

        return SessionKeys(c2sKey: expand("iCam/v1 c2s key", 32),
                           s2cKey: expand("iCam/v1 s2c key", 32),
                           c2sSalt: expand("iCam/v1 c2s salt", 4),
                           s2cSalt: expand("iCam/v1 s2c salt", 4),
                           pairingDigits: String(format: "%06d", value))
    }
}

enum SecureChannelError: Error, Equatable {
    case counterOutOfOrder(expected: UInt64, got: UInt64)
    case decryptionFailed
    case notEstablished
}

/// AES-256-GCM record layer — `docs/PROTOCOL.md` section 4.4.
///
/// One instance per direction. The frame counter is shared across all channels
/// in that direction, so a reordered or replayed frame is detected immediately.
final class SecureChannel {
    private let sealKey: SymmetricKey
    private let openKey: SymmetricKey
    private let sealSalt: Data
    private let openSalt: Data

    private var sendCounter: UInt64 = 0
    private var receiveCounter: UInt64 = 0

    /// `role` decides which derived key seals and which opens.
    init(keys: SessionKeys, role: Role) {
        switch role {
        case .initiator:
            sealKey = SymmetricKey(data: keys.c2sKey)
            openKey = SymmetricKey(data: keys.s2cKey)
            sealSalt = keys.c2sSalt
            openSalt = keys.s2cSalt
        case .responder:
            sealKey = SymmetricKey(data: keys.s2cKey)
            openKey = SymmetricKey(data: keys.c2sKey)
            sealSalt = keys.s2cSalt
            openSalt = keys.c2sSalt
        }
    }

    enum Role: Sendable { case initiator, responder }

    /// 4-byte direction salt followed by the big-endian frame counter.
    /// Throws rather than force-unwrapping: this sits on the media path, and a
    /// crash there would take a live recording with it.
    private static func nonce(salt: Data, counter: UInt64) throws -> AES.GCM.Nonce {
        var raw = salt
        raw.appendUInt64BE(counter)
        do {
            return try AES.GCM.Nonce(data: raw)
        } catch {
            throw SecureChannelError.notEstablished
        }
    }

    /// Seals a payload and returns the complete frame, header included.
    func seal(channel: Channel, flags: FrameFlags = .endOfMessage, plaintext: Data) throws -> Data {
        // The header must be known before sealing because it is the associated
        // data, and its length field already accounts for the 16-byte tag.
        let header = Frame.header(channel: channel, flags: flags,
                                  payloadCount: plaintext.count + 16)
        let box = try AES.GCM.seal(plaintext,
                                   using: sealKey,
                                   nonce: try Self.nonce(salt: sealSalt, counter: sendCounter),
                                   authenticating: header)
        sendCounter &+= 1
        var out = header
        out.append(box.ciphertext)
        out.append(box.tag)
        return out
    }

    /// Opens a received frame. `header` is the raw 8 bytes as they arrived.
    func open(header: Data, payload: Data) throws -> Data {
        guard payload.count >= 16 else { throw SecureChannelError.decryptionFailed }
        let ciphertext = payload.prefix(payload.count - 16)
        let tag = payload.suffix(16)
        let box = try AES.GCM.SealedBox(nonce: try Self.nonce(salt: openSalt, counter: receiveCounter),
                                        ciphertext: ciphertext,
                                        tag: tag)
        do {
            let plaintext = try AES.GCM.open(box, using: openKey, authenticating: header)
            receiveCounter &+= 1
            return plaintext
        } catch {
            throw SecureChannelError.decryptionFailed
        }
    }

    var framesSent: UInt64 { sendCounter }
    var framesReceived: UInt64 { receiveCounter }
}

File: windows/iCam.Core/Security/SecureChannel.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;
using ICam.Core.Protocol;

namespace ICam.Core.Security;

/// <summary>Derived key material for one connection.</summary>
public sealed record SessionKeys(
    byte[] ClientToServerKey,
    byte[] ServerToClientKey,
    byte[] ClientToServerSalt,
    byte[] ServerToClientSalt,
    string PairingDigits)
{
    /// <summary>
    /// The key schedule from <c>docs/PROTOCOL.md</c> section 4.2. Every
    /// primitive here exists natively in both CryptoKit and .NET, which is why
    /// the two implementations need no shared library and no third-party
    /// dependency to agree.
    /// </summary>
    public static SessionKeys Derive(byte[] sharedSecret, byte[] clientRandom, byte[] serverRandom)
    {
        var salt = new byte[clientRandom.Length + serverRandom.Length];
        clientRandom.CopyTo(salt, 0);
        serverRandom.CopyTo(salt, clientRandom.Length);

        var prk = HKDF.Extract(HashAlgorithmName.SHA256, sharedSecret, salt);

        byte[] Expand(string info, int length) =>
            HKDF.Expand(HashAlgorithmName.SHA256, prk, length, Encoding.UTF8.GetBytes(info));

        var sasSeed = Expand("iCam/v1 sas", 8);
        var value = BinaryPrimitives.ReadUInt64BigEndian(sasSeed) % 1_000_000;

        return new SessionKeys(
            Expand("iCam/v1 c2s key", 32),
            Expand("iCam/v1 s2c key", 32),
            Expand("iCam/v1 c2s salt", 4),
            Expand("iCam/v1 s2c salt", 4),
            value.ToString("D6"));
    }
}

public enum ChannelRole
{
    /// <summary>The iPhone. Connects, and proves itself first.</summary>
    Initiator,
    /// <summary>The PC. Listens.</summary>
    Responder,
}

public sealed class SecureChannelException : Exception
{
    public SecureChannelException(string message) : base(message) { }
}

/// <summary>
/// AES-256-GCM record layer — <c>docs/PROTOCOL.md</c> section 4.4.
///
/// One instance per connection. The frame counter is shared across all channels
/// in each direction, so a reordered or replayed frame is detected immediately.
/// </summary>
public sealed class SecureChannel : IDisposable
{
    private readonly AesGcm _seal;
    private readonly AesGcm _open;
    private readonly byte[] _sealSalt;
    private readonly byte[] _openSalt;

    private ulong _sendCounter;
    private ulong _receiveCounter;

    private const int TagSize = 16;
    private const int NonceSize = 12;

    public SecureChannel(SessionKeys keys, ChannelRole role)
    {
        if (role == ChannelRole.Responder)
        {
            _seal = new AesGcm(keys.ServerToClientKey, TagSize);
            _open = new AesGcm(keys.ClientToServerKey, TagSize);
            _sealSalt = keys.ServerToClientSalt;
            _openSalt = keys.ClientToServerSalt;
        }
        else
        {
            _seal = new AesGcm(keys.ClientToServerKey, TagSize);
            _open = new AesGcm(keys.ServerToClientKey, TagSize);
            _sealSalt = keys.ClientToServerSalt;
            _openSalt = keys.ServerToClientSalt;
        }
    }

    public ulong FramesSent => _sendCounter;
    public ulong FramesReceived => _receiveCounter;

    private static void WriteNonce(Span<byte> nonce, ReadOnlySpan<byte> salt, ulong counter)
    {
        salt.CopyTo(nonce);
        BinaryPrimitives.WriteUInt64BigEndian(nonce[salt.Length..], counter);
    }

    /// <summary>Seals a payload and returns the complete frame, header included.</summary>
    public byte[] Seal(Channel channel, ReadOnlySpan<byte> plaintext,
                       FrameFlags flags = FrameFlags.EndOfMessage)
    {
        // The header has to exist before sealing, because it is the associated
        // data — and its length field already accounts for the tag.
        var header = Frame.Header(channel, flags, plaintext.Length + TagSize);

        var result = new byte[header.Length + plaintext.Length + TagSize];
        header.CopyTo(result, 0);

        Span<byte> nonce = stackalloc byte[NonceSize];
        WriteNonce(nonce, _sealSalt, _sendCounter);

        _seal.Encrypt(nonce, plaintext,
                      result.AsSpan(header.Length, plaintext.Length),
                      result.AsSpan(header.Length + plaintext.Length, TagSize),
                      header);
        _sendCounter++;
        return result;
    }

    /// <summary>Opens a received frame. <paramref name="header"/> is the raw eight bytes.</summary>
    public byte[] Open(ReadOnlySpan<byte> header, ReadOnlySpan<byte> payload)
    {
        if (payload.Length < TagSize)
        {
            throw new SecureChannelException("frame is too short to contain an authentication tag");
        }

        var ciphertext = payload[..^TagSize];
        var tag = payload[^TagSize..];
        var plaintext = new byte[ciphertext.Length];

        Span<byte> nonce = stackalloc byte[NonceSize];
        WriteNonce(nonce, _openSalt, _receiveCounter);

        try
        {
            _open.Decrypt(nonce, ciphertext, tag, plaintext, header);
        }
        catch (CryptographicException)
        {
            // A frame that fails authentication means the stream is no longer
            // trustworthy. There is no safe way to skip it and continue.
            throw new SecureChannelException("frame failed authentication");
        }

        _receiveCounter++;
        return plaintext;
    }

    public void Dispose()
    {
        _seal.Dispose();
        _open.Dispose();
    }
}
